src/main/java/org/opensearch/ad/cluster/HashRing.java

```diff
@@ -300,9 +316,19 @@ public void buildCirclesOnAdVersions(Set<String> removedNodeIds, Set<String> add
     private void rebuildAdVersionCirclesWithCooldown() {
         // Check if it's eligible to rebuild hash ring with cooldown
         if (eligibleToRebuildHashRingWithCooldown()) {
-            LOG.info("Rebuild AD version hash ring with cooldown");
+            LOG.info("Rebuild AD version hash ring with cooldown, nodeChangeEvents size {}", nodeChangeEvents.size());
+            int size = nodeChangeEvents.size();
             adVersionCirclesWithCooldown = new TreeMap<>(adVersionCircles);
-            nodeChangeEvents.poll();
+            // It's possible that multiple threads add new event to nodeChangeEvents,
```

Comment:
The copy in line 321 is not a complete deep copy since adVersionCircles is a map of map. Check the following example, where s2 does a copy like yours and s3 is a recursive copy.

```
jshell> TreeMap<String, TreeMap<Integer, Object>> s = new TreeMap<>()
s ==> {}

jshell> TreeMap<Integer, Object> insert = new TreeMap<>()
insert ==> {}

jshell> insert.put(1, new Object())
$3 ==> null

jshell> s.put("1", insert)
$4 ==> null

jshell> TreeMap<String, TreeMap<Integer, Object>> s2 = new TreeMap<>(s)
s2 ==> {1={1=java.lang.Object@17f052a3}}

jshell> s2
s2 ==> {1={1=java.lang.Object@17f052a3}}

jshell> s
s ==> {1={1=java.lang.Object@17f052a3}}

jshell> insert.put(2, new Object())
$9 ==> null

jshell> s
s ==> {1={1=java.lang.Object@17f052a3, 2=java.lang.Object@2e5d6d97}}

jshell> s2
s2 ==> {1={1=java.lang.Object@17f052a3, 2=java.lang.Object@2e5d6d97}}

jshell> for (Map.Entry<String, TreeMap<Integer, Object>> entry: s.entrySet()) {
   ...>   s3.put(new String(entry.getKey()), new TreeMap<>(entry.getValue()));
   ...> }

jshell> s3
s3 ==> {1={1=java.lang.Object@17f052a3, 2=java.lang.Object@2e5d6d97}}

jshell> s
s ==> {1={1=java.lang.Object@17f052a3, 2=java.lang.Object@2e5d6d97}}

jshell> insert.put(3, new Object())
$16 ==> null

jshell> s
s ==> {1={1=java.lang.Object@17f052a3, 2=java.lang.Object@2e5d6d97, 3=java.lang.Object@6267c3bb}}

jshell> s3
s3 ==> {1={1=java.lang.Object@17f052a3, 2=java.lang.Object@2e5d6d97}}
```

Reply:
Good point, will fix